### PR TITLE
Add Rannacher-Turek element.

### PR DESCRIPTION
**Pull request by pesser**
Lowest order Rannacher Turek polynomials and FE in 2D and tests.

### CODE REVIEW

File: include/deal.II/base/polynomials_rannacher_turek.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * Basis for polynomial space on the unit square used for lowest order
+ * Rannacher Turek element.
+ *
+ * The i-th basis function is the dual basis element corresponding to
+ * the dof which evaluates the functions mean value across the i-th
```
Comment (bangerth) [inline include/deal.II/base/polynomials_rannacher_turek.h]: functions -> function's

File: include/deal.II/fe/fe_rannacher_turek.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+   */
+  FE_RannacherTurek(const unsigned int degree = 0,
+                    const unsigned int n_face_support_points = 2);
+
+  virtual std::string get_name() const;
+  virtual FiniteElement<dim> *clone() const;
+
+  virtual void interpolate(
+    std::vector<double> &local_dofs,
+    const std::vector<double> &values) const;
```
Comment (bangerth) [inline include/deal.II/fe/fe_rannacher_turek.h]: Is this a function you needed to implement for the interface of one of the base classes? If not, some documentation would be good. Where do you need it and what does it do?

File: source/base/polynomials_rannacher_turek.cc [hunk 3]
```diff
@@ -0,0 +1,156 @@
+#include <deal.II/base/polynomials_rannacher_turek.h>
```
Comment (bangerth) [inline source/base/polynomials_rannacher_turek.cc]: Can you add the standard header to this file (as well as the others)? Among other things this indicates that you agree to releasing your code under the LGPL. We need this agreement to distribute your code.

File: source/base/polynomials_rannacher_turek.inst.in [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1998 - 2014 by the deal.II authors
```
Comment (bangerth) [inline source/base/polynomials_rannacher_turek.inst.in]: Just list 2015 here.

File: source/fe/fe_rannacher_turek.cc [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+template <int dim>
+FE_RannacherTurek<dim>::FE_RannacherTurek(const unsigned int degree,
+                                          const unsigned int n_face_support_points) :
+  FE_Poly<PolynomialsRannacherTurek<dim>, dim>(
+    PolynomialsRannacherTurek<dim>(),
+    FiniteElementData<dim>(this->get_dpo_vector(),
+                           1,
+                           2,
+                           FiniteElementData<dim>::L2),
```
Comment (bangerth) [inline source/fe/fe_rannacher_turek.cc]: I forgot -- is the Rannacher-Turek element really completely discontinuous, or is it at least H_div conforming?

File: source/fe/fe_rannacher_turek.cc [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  const std::vector<double> &values) const
+{
+  AssertDimension(values.size(), this->generalized_support_points.size());
+  AssertDimension(local_dofs.size(), this->dofs_per_cell);
+
+  std::size_t q_points_per_face = this->weights.size();
+  std::fill(local_dofs.begin(), local_dofs.end(), 0.0);
+
+  typedef std::vector<double>::const_iterator it;
+  it value = values.begin();
```
Comment (bangerth) [inline source/fe/fe_rannacher_turek.cc]: You only use the typedef once. Let's just spell things out to make code easier to read.

File: source/fe/fe_rannacher_turek.cc [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+template <int dim>
+void FE_RannacherTurek<dim>::interpolate(
+  std::vector<double> &local_dofs,
+  const std::vector<double> &values) const
+{
+  AssertDimension(values.size(), this->generalized_support_points.size());
+  AssertDimension(local_dofs.size(), this->dofs_per_cell);
+
+  std::size_t q_points_per_face = this->weights.size();
```
Comment (bangerth) [inline source/fe/fe_rannacher_turek.cc]: Just use `const unsigned int` for the type here.

**Comment by bangerth** [comment]
Nice work, thank you very much! I especially appreciate the fact that you added tests as well!

I only had a few small comments. Can you address them and add an entry to `doc/news/changes.h` to the same place where the `FE_QBubbbles` by Daniel Arndt is already discussed?

File: include/deal.II/fe/fe_rannacher_turek.h (hunk 2 above)
Comment (pesser) [inline include/deal.II/fe/fe_rannacher_turek.h]: `interpolate` is in the interface of the base FiniteElement class so I let doxygen take the documentation found there.

File: source/base/polynomials_rannacher_turek.cc (hunk 3 above)
Comment (pesser) [inline source/base/polynomials_rannacher_turek.cc]: done.

File: source/fe/fe_rannacher_turek.cc (hunk 5 above)
Comment (pesser) [inline source/fe/fe_rannacher_turek.cc]: It is scalar valued so I think H_div conforming would be confusing anyway but even if we use a system with Rannacher-Turek components the property that jumps across faces are mean value free is not enough for H_div conformity (AFAIK you need continuity of normal moments w.r.t. all trace functions, not just the constant ones)

File: include/deal.II/fe/fe_rannacher_turek.h (hunk 2 above)
Comment (bangerth) [inline include/deal.II/fe/fe_rannacher_turek.h]: Ah, fair enough. Another rather poorly named function in one of our base classes :-(

File: source/fe/fe_rannacher_turek.cc (hunk 5 above)
Comment (bangerth) [inline source/fe/fe_rannacher_turek.cc]: OK, thanks.

**Comment by pesser** [comment]
Thanks for the feedback. I've tried to update everything accordingly.

**Comment by bangerth** [comment]
Great, I'm very excited to see the element of two of my former advisers finally implemented :-)

Thanks for all the work. I hope it won't remain your last contribution!

**Pull request merged**